File: frontend/src/services/authService.ts
import api from './api';
import { ApiResponse, LoginRequest, RegisterRequest, TokenResponse } from '../types';

/**
 * Serviço de autenticação - Facade Pattern
 */
export const authService = {
  /**
   * Realiza login do usuário
   */
  async login(credentials: LoginRequest): Promise<TokenResponse> {
    const response = await api.post<ApiResponse<TokenResponse>>(
      '/api/v1/auth/login',
      credentials
    );
    const data = response.data.data;
    
    // Armazenar tokens
    localStorage.setItem('accessToken', data.accessToken);
    localStorage.setItem('refreshToken', data.refreshToken);
    localStorage.setItem('user', JSON.stringify(data.user));
    
    return data;
  },

  /**
   * Registra novo usuário
   */
  async register(userData: RegisterRequest): Promise<TokenResponse> {
    const response = await api.post<ApiResponse<TokenResponse>>(
      '/api/v1/auth/register',
      userData
    );
    const data = response.data.data;
    
    // Armazenar tokens
    localStorage.setItem('accessToken', data.accessToken);
    localStorage.setItem('refreshToken', data.refreshToken);
    localStorage.setItem('user', JSON.stringify(data.user));
    
    return data;
  },

  /**
   * Renova o token de acesso
   */
  async refreshToken(): Promise<TokenResponse> {
    const refreshToken = localStorage.getItem('refreshToken');
    if (!refreshToken) {
      throw new Error('Refresh token não encontrado');
    }

    const response = await api.post<ApiResponse<TokenResponse>>(
      '/api/v1/auth/refresh',
      { refreshToken }
    );
    const data = response.data.data;
    
    // Atualizar tokens
    localStorage.setItem('accessToken', data.accessToken);
    localStorage.setItem('refreshToken', data.refreshToken);
    
    return data;
  },

  /**
   * Realiza logout do usuário
   */
  logout(): void {
    localStorage.removeItem('accessToken');
    localStorage.removeItem('refreshToken');
    localStorage.removeItem('user');
  },

  /**
   * Verifica se o usuário está autenticado
   */
  isAuthenticated(): boolean {
    return !!localStorage.getItem('accessToken');
  },

  /**
   * Obtém o usuário atual
   */
  getCurrentUser() {
    const user = localStorage.getItem('user');
    return user ? JSON.parse(user) : null;
  },

  /**
   * Obtém o token de acesso
   */
  getAccessToken(): string | null {
    return localStorage.getItem('accessToken');
  },
};

File: backend/src/main/java/com/artistalbum/service/AuthService.java
package com.artistalbum.service;

import com.artistalbum.dto.AuthDTO;
import com.artistalbum.entity.User;
import com.artistalbum.exception.BusinessException;
import com.artistalbum.security.JwtService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.stereotype.Service;

/**
 * Serviço de autenticação JWT.
 */
@Service
@RequiredArgsConstructor
@Slf4j
public class AuthService {

    private final AuthenticationManager authenticationManager;
    private final JwtService jwtService;
    private final UserService userService;

    /**
     * Realiza login e retorna tokens JWT.
     */
    public AuthDTO.TokenResponse login(AuthDTO.LoginRequest request) {
        log.info("Tentativa de login para usuário: {}", request.getUsername());

        Authentication authentication = authenticationManager.authenticate(
                new UsernamePasswordAuthenticationToken(
                        request.getUsername(),
                        request.getPassword()
                )
        );

        User user = (User) authentication.getPrincipal();
        return generateTokenResponse(user);
    }

    /**
     * Registra novo usuário e retorna tokens JWT.
     */
    public AuthDTO.TokenResponse register(AuthDTO.RegisterRequest request) {
        log.info("Registrando novo usuário: {}", request.getUsername());

        User user = userService.createUser(
                request.getUsername(),
                request.getEmail(),
                request.getPassword(),
                request.getFullName()
        );

        return generateTokenResponse(user);
    }

    /**
     * Renova o token de acesso usando o refresh token.
     */
    public AuthDTO.TokenResponse refreshToken(String refreshToken) {
        log.debug("Renovando token de acesso");

        try {
            String username = jwtService.extractUsername(refreshToken);
            User user = userService.findByUsername(username);

            if (!jwtService.isTokenValid(refreshToken, user)) {
                throw new BusinessException("Refresh token inválido ou expirado");
            }

            return generateTokenResponse(user);
        } catch (Exception e) {
            log.error("Erro ao renovar token: {}", e.getMessage());
            throw new BusinessException("Não foi possível renovar o token");
        }
    }

    /**
     * Gera resposta com tokens JWT.
     */
    private AuthDTO.TokenResponse generateTokenResponse(User user) {
        String accessToken = jwtService.generateToken(user);
        String refreshToken = jwtService.generateRefreshToken(user);

        return AuthDTO.TokenResponse.builder()
                .accessToken(accessToken)
                .refreshToken(refreshToken)
                .tokenType("Bearer")
                .expiresIn(jwtService.getExpirationTime() / 1000) // em segundos
                .user(AuthDTO.UserInfo.builder()
                        .id(user.getId())
                        .username(user.getUsername())
                        .email(user.getEmail())
                        .fullName(user.getFullName())
                        .role(user.getRole().name())
                        .build())
                .build();
    }
}
